### PR TITLE
[FLINK-23976][metrics] Add additional job status metrics

### PR DESCRIPTION
**Pull request by zentol**
Adds new metrics for observing the job status.

Metrics come in 3 variants for each non-terminal job state, which can be configured by the user:
* "<jobStatus>State" returns whether the job is currently in a state (1=yes, 0=no)
* "<jobStatus>Time" returns how long the job has been in the state since the last state transition
* "<jobStatus>TimeTotal" return how long the job spent in the state in total

The uptime/downtime metrics were deprecated because as is they are not very well-defined.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java [hunk 1]
```diff
@@ -599,20 +608,117 @@ public ExecutionGraph getExecutionGraph() {
...
+                executionGraph,
+                this::getNumberOfRestarts,
+                ignored -> {},
+                executionGraph.getStatusTimestamp(JobStatus.INITIALIZING),
+                jobStatusMetricsSettings);
         operatorCoordinatorHandler.startAllOperatorCoordinators();
         startSchedulingInternal();
     }
 
     public static void registerJobMetrics(
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:622]: Would it make sense to get rid of the `jobStatusListenerConsumer`? I've found it bit confusing at first.
```suggestion
    public static JobStatusListener registerJobStatusMetrics(
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java [hunk 2]
```diff
@@ -599,20 +608,117 @@ public ExecutionGraph getExecutionGraph() {
     @Override
     public final void startScheduling() {
         mainThreadExecutor.assertRunningInMainThread();
-        registerJobMetrics(jobManagerJobMetricGroup, executionGraph, this::getNumberOfRestarts);
+        registerJobMetrics(
+                jobManagerJobMetricGroup,
+                executionGraph,
+                this::getNumberOfRestarts,
+                ignored -> {},
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:615]: How does this work for the default scheduler if we don't listen on status changes?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java [hunk 3]
```diff
@@ -496,8 +495,7 @@ public void testNumRestartsMetric() throws Exception {
     public void testStatusMetrics() throws Exception {
         final CompletableFuture<UpTimeGauge> upTimeMetricFuture = new CompletableFuture<>();
         final CompletableFuture<DownTimeGauge> downTimeMetricFuture = new CompletableFuture<>();
-        final CompletableFuture<RestartTimeGauge> restartTimeMetricFuture =
-                new CompletableFuture<>();
+        new CompletableFuture<>();
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:498]: typo?

File: flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java [hunk 4]
```diff
@@ -215,5 +218,89 @@
                                     + "faster updating metrics. Increase this value if the metric fetcher causes too much load. Setting this value to 0 "
                                     + "disables the metric fetching completely.");
 
+    /** Controls which job status metrics will be exposed. */
+    public static final ConfigOption<List<JobStatusMetrics>> JOB_STATUS_METRICS =
+            key("metrics.job.status.enable")
```
Comment (dmvk) [inline flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java:223]: this is just to reduce cardinality or is there any other reason?

**Review by dmvk — COMMENTED** [review]
Thanks for the PR @zentol, this seems to be headed in a right direction! I've added few questions, PTAL

File: flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java (hunk 4 above)
Comment (zentol) [reply flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java:223]: > to reduce cardinality

yup.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:622]: I'm not too fond of it either, but returning something doesn't scale in the long-term imo. What happens the next time we'd need to add a parameter that isn't quite as nice as we'd like?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:498]: yes

**Review by dmvk — APPROVED** [review]
LGTM 👍 Thanks for addressing the comments.

**Comment by dmvk** [comment]
Can you please squash the commit history before merging?

**Pull request merged**